### PR TITLE
Scala akka-http server

### PR DESCRIPTION
**Pull request by Bouillie**
This is a basic scala akka-http server implementation.

https://github.com/OpenAPITools/openapi-generator/issues/5711

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@clasnake, @jimschubert, @shijinkui, @ramzimaalej, @chameleon82

### CODE REVIEW

**Comment by Bouillie** [comment]
Any idea how to deal with this build error ?
I made no change to kotlin...

```
[ERROR] Incompatible classes were found in dependencies. Remove them from the classpath or use '-Xskip-metadata-version-check' to suppress errors
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/kotlin/vertx/src/main/kotlin/org/openapitools/server/api/verticle/PetApiVertxProxyHandler.kt: (12, 27) Class 'kotlinx.coroutines.GlobalScope' is compiled by a pre-release version of Kotlin and cannot be loaded by this version of the compiler
```

**Comment by wing328** [comment]
Nothing to do with this PR and we've temporarily disabled the failed Kotlin tests for the time being. Please merge the latest master into this branch.

**Comment by Bouillie** [comment]
I will. In the meantime I found some incompatibilities in my generated code with scala 2.13 and some versions of akka-http, I'll fix that too.

**Comment by Bouillie** [comment]
I think the PR is ready.

**Comment by wing328** [comment]
I ran `sbt test` on the petstore sample but got the following errors:
```
$ sbt test
[info] Loading project definition from /Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/project
[info] Loading settings from build.sbt ...
[info] Set current project to openapi-scala-akka-http-server (in build file:/Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/)
[info] Updating ...
[info] downloading https://repo1.maven.org/maven2/com/typesafe/akka/akka-http_2.12/10.1.9/akka-http_2.12-10.1.9.jar ...
[info] downloading https://repo1.maven.org/maven2/com/typesafe/akka/akka-parsing_2.12/10.1.9/akka-parsing_2.12-10.1.9.jar ...
[info] downloading https://repo1.maven.org/maven2/com/typesafe/akka/akka-http-core_2.12/10.1.9/akka-http-core_2.12-10.1.9.jar ...
[info] 	[SUCCESSFUL ] com.typesafe.akka#akka-http_2.12;10.1.9!akka-http_2.12.jar (2122ms)
[info] 	[SUCCESSFUL ] com.typesafe.akka#akka-parsing_2.12;10.1.9!akka-parsing_2.12.jar (2867ms)
[info] 	[SUCCESSFUL ] com.typesafe.akka#akka-http-core_2.12;10.1.9!akka-http-core_2.12.jar (3354ms)
[info] Done updating.
[info] Compiling 13 Scala sources to /Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/target/scala-2.12/classes ...
[error] /Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/src/main/scala/org/openapitools/server/StringDirectives.scala:65:120: too many arguments (2) for method apply: (supported: scala.collection.immutable.Set[akka.http.scaladsl.model.ContentTypeRange])akka.http.scaladsl.server.UnsupportedRequestContentTypeRejection in object UnsupportedRequestContentTypeRejection
[error]       case Failure(x: UnsupportedContentTypeException) => reject(UnsupportedRequestContentTypeRejection(x.supported, x.actualContentType))
[error]                                                                                                                        ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 28 s, completed Apr 3, 2020 11:05:58 PM
```

Does it work for you locally in your machine?

**Comment by Bouillie** [comment]
I made a mistake on the akka-http version I has to check. It's fixed now.

**Comment by wing328** [comment]
Tested again and the result is good:
```
[info] Loading project definition from /Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/project
[info] Loading settings from build.sbt ...
[info] Set current project to openapi-scala-akka-http-server (in build file:/Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/)
[info] Updating ...
[info] downloading https://repo1.maven.org/maven2/com/typesafe/akka/akka-parsing_2.12/10.1.10/akka-parsing_2.12-10.1.10.jar ...
[info] downloading https://repo1.maven.org/maven2/com/typesafe/akka/akka-http_2.12/10.1.10/akka-http_2.12-10.1.10.jar ...
[info] downloading https://repo1.maven.org/maven2/com/typesafe/akka/akka-http-core_2.12/10.1.10/akka-http-core_2.12-10.1.10.jar ...
[info] 	[SUCCESSFUL ] com.typesafe.akka#akka-parsing_2.12;10.1.10!akka-parsing_2.12.jar (2649ms)
[info] 	[SUCCESSFUL ] com.typesafe.akka#akka-http_2.12;10.1.10!akka-http_2.12.jar (2647ms)
[info] 	[SUCCESSFUL ] com.typesafe.akka#akka-http-core_2.12;10.1.10!akka-http-core_2.12.jar (4247ms)
[info] Done updating.
[info] Compiling 13 Scala sources to /Users/williamcheng/Code/openapi-generator/samples/server/petstore/scala-akka-http/target/scala-2.12/classes ...
[warn] there were three feature warnings; re-run with -feature for details
[warn] one warning found
[info] Done compiling.
[success] Total time: 79 s, completed Apr 4, 2020 2:42:29 PM

```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+    public static final String DEFAULT_AKKA_HTTP_VERSION = "10.1.10";
+
+    static Logger LOGGER = LoggerFactory.getLogger(ScalaAkkaHttpServerCodegen.class);
+
+    public CodegenType getTag() {
+        return CodegenType.SERVER;
+    }
+
+    public String getName() {
+        return "scala-akka-http";
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java:36]: I'll rename it to "scala-akka-http-server" to follow our generator naming convention.

**Comment by wing328** [comment]
@Bouillie thanks for the PR, which has been merged into master.

I'll file a PR with minor improvements to the generator.

Have a nice weekend.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+
+    @Override
+    public CodegenParameter fromParameter(Parameter parameter, Set<String> imports) {
+        CodegenParameter param = super.fromParameter(parameter, imports);
+        // Removing unhandled types
+        if(!primitiveParamTypes.contains(param.dataType)) {
+            param.dataType = "String";
+        }
+        if (!param.required) {
+            param.vendorExtensions.put("hasDefaultValue", param.defaultValue != null);
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java:233]: We've trying to normalize the naming of vendor extensions. I wonder if you can rename `hasDefaultValue` to `x-has-default-value` instead.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+                }
+            } else {
+                textOrMatcher.value = path;
+                textOrMatcher.isText = true;
+                pathMatchers.add(textOrMatcher);
+            }
+        }
+        pathMatchers.getLast().hasMore = false;
+
+        codegenOperation.vendorExtensions.put("paths", pathMatchers);
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java:307]: We've trying to normalize the naming of vendor extensions. I wonder if you can rename `paths` to `x-paths` instead.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+                }
+
+                HashSet<Marshaller> operationSpecificMarshallers = new HashSet<>();
+                for (CodegenResponse response : op.responses) {
+                    if (!response.primitiveType) {
+                        Marshaller marshaller = new Marshaller(response);
+                        entityMarshallerTypes.add(marshaller);
+                        operationSpecificMarshallers.add(marshaller);
+                    }
+                    response.vendorExtensions.put("isDefault", response.code.equals("0"));
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java:392]: We've trying to normalize the naming of vendor extensions. I wonder if you can rename `x-isDefault` to `x-is-default` instead.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+                }
+                op.vendorExtensions.put("specificMarshallers", operationSpecificMarshallers);
+                op.vendorExtensions.put("fileParams", fileParams);
+                op.vendorExtensions.put("nonFileParams", nonFileParams);
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ScalaAkkaHttpServerCodegen.java:396]: We're trying to normalize the naming of vendor extensions. I wonder if you can rename the extensions as followings:
- `specificMarshallers ` to `x-specific-marshallers `
- `fileParams` to `x-file-params`
- `nonFileParams` to `x-non-file-params`

**Pull request merged**